pkg/util/cluster/cluster.go

```diff
@@ -346,20 +357,44 @@ func (c *Cluster) generateSubnets() (vnetPrefix string, masterSubnet string, wor
 	// 10.0.0.0/16 is used by dev-vnet to host CI
 	// 10.1.0.0/24 is used by rp-vnet to host Proxy VM
 	// 10.2.0.0/24 is used by dev-vpn-vnet to host VirtualNetworkGateway
-	var x, y int
 	rand.Seed(time.Now().UnixNano())
-	// Local Dev clusters are limited to /16 dev-vnet
-	if !c.ci {
-		x, y = 0, 2*rand.Intn(128)
-	} else {
-		x, y = rand.Intn((124))+3, 2*rand.Intn(128)
-	}
+	x, y := rand.Intn((124))+3, 2*rand.Intn(128)
 	vnetPrefix = fmt.Sprintf("10.%d.%d.0/23", x, y)
 	masterSubnet = fmt.Sprintf("10.%d.%d.0/24", x, y)
 	workerSubnet = fmt.Sprintf("10.%d.%d.0/24", x, y+1)
 	return
 }
 
+func (c *Cluster) getUnusedAddressSpaces(ctx context.Context, vnetResourceGroup string) (string, string, string, error) {
+	addressSpaces := make([]string, 0)
+	subnetListResultPage, err := c.subnets.List(ctx, vnetResourceGroup, devVnetName)
+	if err != nil {
+		return "", "", "", err
+	}
+	mp := make(map[string]bool)
+	for _, subnet := range subnetListResultPage.Values() {
+		mp[*subnet.AddressPrefix] = true
+	}
+
+	devVnet, err := c.vnets.Get(ctx, c.env.ResourceGroup(), devVnetName, "")
+	if err != nil {
+		return "", "", "", err
+	}
+
+	for _, addressPrefix := range *devVnet.AddressSpace.AddressPrefixes {
+		if mp[addressPrefix] {
+			continue
+		}
+		addressSpaces = append(addressSpaces, addressPrefix)
+	}
+
+	if len(addressSpaces) < 2 {
+		return "", "", "", fmt.Errorf("not enough address space available in the virtual network %s", devVnetName)
+	}
+	addressPrefix := addressSpaces[0][:len(addressSpaces[0])-1] + "3"
```

Comment:
the algo here seems not very right.
IIUC, for the addressPrefix, it simply picks up the second last char and append a "3" to it, however this only works for /24 subnet. it breaks when subnet mask length changes to other value, eg, 23 or 25, (not sure if this may happen tho). I would suggest to change this to something like:

```
	addr := strings.Split(addressSpaces[0], "/")
	i, _ := strconv.Atoi(addr[1])
	addressPrefix = addr[0] + "/" + strconv.Itoa(i-1)
```

Also, I believe we need to make sure that subnets listed here are consecutive, otherwise the first and second available subnets are probably not merge-able and then the Prefix is invalid. probably need a better algo to calculate it. :thinking:

Reply:
the lines to comment is a bit messed up, should be this:

```
addressPrefix := addressSpaces[0][:len(addressSpaces[0])-1] + "3"
return addressPrefix, addressSpaces[0], addressSpaces[1], nil
```

Reply:
Yes, thank you @yjst2012 , you are right, I will make the changes.